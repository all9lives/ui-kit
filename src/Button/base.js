import { css } from 'styled-components'
import Color from '../utils/Color'
import Unit from '../utils/Unit'
import Size from '../utils/Size'

export const attrs = {
  color: props => props.primary
    ? Color.primary
    : (Color[props.color] || props.color || Color.black)
}

const base = css`
  margin: ${props => Unit.parse(props.margin, (margin) => margin * Size.spacing)};
`

export default base
