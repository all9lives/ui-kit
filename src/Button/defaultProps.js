export default {
  title: 'Button',
  onPress: () => {},
  margin: 0
}
