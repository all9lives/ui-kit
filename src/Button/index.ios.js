import Button, { StyledText, StyledView } from './index.android'

export { StyledText }
export { StyledView }
export default Button
