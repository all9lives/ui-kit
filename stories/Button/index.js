import React from 'react'
import { action } from '@storybook/addon-actions'
import createStory from '../../storybook/createStory'
import Button from '../../src/Button'

createStory('Button', module)
  .add('with `children:Apple`', () => <Button>Apple</Button>)
  .add('with `title:Banana`', () => <Button title='Banana' />)
  .add('with `onPress`', () =>
    <Button
      title='Press me'
      onPress={action('Pressed')}
    />
  )
  .add('with `disable:true`', () =>
    <Button disabled />
  )
  .add('with `primary`', () =>
    <Button primary />
  )
  .add('with `color:success`', () =>
    <Button color='success' />
  )
  .add('with `color:white`', () =>
    <Button color='white' />
  )
  .add('with `margin:2`', () =>
    <Button margin={2} />
  )
  .add('with `margin:2 1`', () =>
    <Button margin='2 1' />
  )
  .add('with `style`', () =>
    <Button style={{ width: 200, alignItems: 'center' }} />
  )
  .add('with `padding`', () =>
    <Button padding='2 3' />
  )
