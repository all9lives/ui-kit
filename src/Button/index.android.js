import React from 'react'
import { TouchableOpacity } from 'react-native'
import styled from 'styled-components/native'
import tinycolor from 'tinycolor2'
import { attrs } from './base'
import defaultProps from './defaultProps'
import View from '../View'
import Color from '../utils/Color'
import Size from '../utils/Size'

export const StyledView = styled(View).attrs({
  ...attrs
})`
  background-color: ${props => props.color};
  padding: ${Size.spacing}px ${Size.spacing * 3}px;
  border-radius: ${Size.spacing * 2};
`
StyledView.defaultProps = {
  margin: 0
}

export const StyledText = styled.Text.attrs({
  ...attrs
})`
  color: ${props => Color[tinycolor(props.color).getBrightness() < 200 ? 'white' : 'black']};
`

const Button = ({ onPress, disabled, margin, style, children, title, ...props }) => (
  <TouchableOpacity disabled={disabled} onPress={onPress}>
    <StyledView {...props} margin={margin} style={{ opacity: disabled ? 0.4 : 1 }}>
      <StyledText {...props}>
        {children || title}
      </StyledText>
    </StyledView>
  </TouchableOpacity>
)
Button.defaultProps = defaultProps

export default Button
