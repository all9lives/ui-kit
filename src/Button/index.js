import styled from 'styled-components'
import { invert } from '../utils/Color'
import base, { attrs } from './base'
import defaultProps from './defaultProps'

const Button = styled.button.attrs({
  ...attrs,
  children: props => (props.children || props.title),
  onClick: props => (props.onClick || props.onPress)
})`
  ${base}
  background: none;
  border: 0;
  font: inherit;
  line-height: normal;
  overflow: visible;
  user-select: none;
  padding: 0.3rem 1rem;
  border-radius: 1rem;
  color: ${props => invert(props.color, true)};
  background-color: ${props => props.color};
  opacity: ${props => props.disabled ? '0.4' : 1};

  :active {
    transform: translate3d(1px, 1px, 0);
  }
`
Button.defaultProps = defaultProps

export default Button
