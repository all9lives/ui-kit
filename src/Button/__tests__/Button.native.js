/* eslint-env jest */
import React from 'react'
import { shallow } from 'enzyme'
import Button, { StyledText, StyledView } from '../index.ios.js'

describe('Button.native', () => {
  it('renders a snapshot', () => {
    const wrapper = shallow(
      <Button>Apple</Button>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('renders a snapshot with `title:Press`', () => {
    const wrapper = shallow(
      <Button title='Press' />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('renders a snapshot with `onPress`', () => {
    const wrapper = shallow(
      <Button onPress={() => {}} />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('renders a snapshot with `disabled`', () => {
    const wrapper = shallow(
      <Button disabled />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('renders a snapshot with `primary`', () => {
    const wrapper = shallow(
      <Button primary />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('renders a snapshot with `color:white`', () => {
    const wrapper = shallow(
      <Button color='white' />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('triggers something when onPressed', () => {
    const wrapper = shallow(
      <Button />
    )
    wrapper.props().onPress()
  })

  it('returns something when onPressed', () => {
    const wrapper = shallow(
      <Button onPress={() => 'onPressed'} />
    )
    const result = wrapper.props().onPress()
    expect(result).toEqual('onPressed')
  })
})

describe('Button.StyledView.native', () => {
  it('renders a snapshot', () => {
    const wrapper = shallow(
      <StyledView />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('renders a snapshot with `color:white`', () => {
    const wrapper = shallow(
      <StyledView color='white' />
    )
    expect(wrapper).toMatchSnapshot()
  })
})

describe('Button.StyledText.native', () => {
  it('renders a snapshot', () => {
    const wrapper = shallow(
      <StyledText>Text</StyledText>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('renders a snapshot with `color:white`', () => {
    const wrapper = shallow(
      <StyledText color='white'>Text</StyledText>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('renders a snapshot with `title:Text`', () => {
    const wrapper = shallow(
      <StyledText title='Text' />
    )
    expect(wrapper).toMatchSnapshot()
  })
})
