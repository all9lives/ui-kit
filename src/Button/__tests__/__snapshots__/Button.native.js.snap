// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button.StyledText.native renders a snapshot 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  color="#191919"
  disabled={false}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "color": "#fafafa",
      },
      undefined,
    ]
  }
>
  Text
</Text>
`;

exports[`Button.StyledText.native renders a snapshot with \`color:white\` 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  color="white"
  disabled={false}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "color": "#191919",
      },
      undefined,
    ]
  }
>
  Text
</Text>
`;

exports[`Button.StyledText.native renders a snapshot with \`title:Text\` 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  color="#191919"
  disabled={false}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "color": "#fafafa",
      },
      undefined,
    ]
  }
  title="Text"
/>
`;

exports[`Button.StyledView.native renders a snapshot 1`] = `
<Styled(View)
  align=""
  color="#191919"
  direction="column"
  innerRef={[Function]}
  margin={0}
  padding={0}
  style={
    Array [
      Object {
        "backgroundColor": "#191919",
        "borderRadius": 20,
        "paddingBottom": 10,
        "paddingLeft": 20,
        "paddingRight": 20,
        "paddingTop": 10,
      },
      undefined,
    ]
  }
/>
`;

exports[`Button.StyledView.native renders a snapshot with \`color:white\` 1`] = `
<Styled(View)
  align=""
  color="white"
  direction="column"
  innerRef={[Function]}
  margin={0}
  padding={0}
  style={
    Array [
      Object {
        "backgroundColor": "#fafafa",
        "borderRadius": 20,
        "paddingBottom": 10,
        "paddingLeft": 20,
        "paddingRight": 20,
        "paddingTop": 10,
      },
      undefined,
    ]
  }
/>
`;

exports[`Button.native renders a snapshot 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  onPress={[Function]}
>
  <Styled(Styled(View))
    margin={0}
    style={
      Array [
        undefined,
        Object {
          "opacity": 1,
        },
      ]
    }
  >
    <Styled(Text)>
      Apple
    </Styled(Text)>
  </Styled(Styled(View))>
</TouchableOpacity>
`;

exports[`Button.native renders a snapshot with \`color:white\` 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  onPress={[Function]}
>
  <Styled(Styled(View))
    color="white"
    margin={0}
    style={
      Array [
        undefined,
        Object {
          "opacity": 1,
        },
      ]
    }
  >
    <Styled(Text)
      color="white"
    >
      Button
    </Styled(Text)>
  </Styled(Styled(View))>
</TouchableOpacity>
`;

exports[`Button.native renders a snapshot with \`disabled\` 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  disabled={true}
  onPress={[Function]}
>
  <Styled(Styled(View))
    margin={0}
    style={
      Array [
        undefined,
        Object {
          "opacity": 0.4,
        },
      ]
    }
  >
    <Styled(Text)>
      Button
    </Styled(Text)>
  </Styled(Styled(View))>
</TouchableOpacity>
`;

exports[`Button.native renders a snapshot with \`onPress\` 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  onPress={[Function]}
>
  <Styled(Styled(View))
    margin={0}
    style={
      Array [
        undefined,
        Object {
          "opacity": 1,
        },
      ]
    }
  >
    <Styled(Text)>
      Button
    </Styled(Text)>
  </Styled(Styled(View))>
</TouchableOpacity>
`;

exports[`Button.native renders a snapshot with \`padding:2\` 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  onPress={[Function]}
>
  <Styled(Styled(View))
    margin={0}
    style={
      Array [
        undefined,
        Object {
          "opacity": 1,
        },
      ]
    }
  >
    <Styled(Text)>
      Text
    </Styled(Text)>
  </Styled(Styled(View))>
</TouchableOpacity>
`;

exports[`Button.native renders a snapshot with \`primary\` 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  onPress={[Function]}
>
  <Styled(Styled(View))
    margin={0}
    primary={true}
    style={
      Array [
        undefined,
        Object {
          "opacity": 1,
        },
      ]
    }
  >
    <Styled(Text)
      primary={true}
    >
      Button
    </Styled(Text)>
  </Styled(Styled(View))>
</TouchableOpacity>
`;

exports[`Button.native renders a snapshot with \`style\` 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  onPress={[Function]}
>
  <Styled(Styled(View))
    margin={0}
    style={
      Array [
        Object {
          "margin": 10,
        },
        Object {
          "opacity": 1,
        },
      ]
    }
  >
    <Styled(Text)>
      Button
    </Styled(Text)>
  </Styled(Styled(View))>
</TouchableOpacity>
`;

exports[`Button.native renders a snapshot with \`title:Press\` 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  onPress={[Function]}
>
  <Styled(Styled(View))
    margin={0}
    style={
      Array [
        undefined,
        Object {
          "opacity": 1,
        },
      ]
    }
  >
    <Styled(Text)>
      Press
    </Styled(Text)>
  </Styled(Styled(View))>
</TouchableOpacity>
`;
